### PR TITLE
Universal adapter kotlin

### PR DESCRIPTION
**Pull request by marcin-adamczewski**
Note that this PR is based on another PR

### CODE REVIEW

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+     * Keep in mind that the View in question is already removed from the RecyclerView.
+     *
+     * @see RecyclerView.Adapter.onFailedToRecycleView
+     * @return True if the View should be recycled, false otherwise. Note that if this method
+     * returns `true`, RecyclerView *will ignore* the transient state of
+     * the View and recycle it regardless. If this method returns `false`,
+     * RecyclerView will check the View's transient state again before giving a final decision.
+     * Default implementation returns false.
+     */
+    fun onFailedToRecycleView(): Boolean {
```
Comment (jacek-marchwicki) [inline universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt]: I think those should be `open`

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        return false
+    }
+
+    /**
+     * Called when a view created by this view holder has been attached to a window.
+     *
+     * @see RecyclerView.Adapter.onViewAttachedToWindow
+     */
+    fun onViewAttachedToWindow() {}
```
Comment (jacek-marchwicki) [inline universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt]: I think those should be `open`

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * Called when a view created by this view holder has been detached from its window.
+     *
+     * @see RecyclerView.Adapter.onViewDetachedFromWindow
+     */
+    fun onViewDetachedFromWindow() {}
```
Comment (jacek-marchwicki) [inline universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt]: I think those should be `open`

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolderManager.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.annotation.LayoutRes
+import com.jacekmarchwicki.universaladapter.BaseAdapterItem
+import kotlin.reflect.KClass
+
+open class BaseViewHolderManager<T : BaseAdapterItem>(
+        @LayoutRes private val layoutRes: Int,
```
Comment (jacek-marchwicki) [inline universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolderManager.kt]: I wonder if this a good approach to always requiring `layoutRes`. What if someone doesn't want to use an XML file, and wants just creating the view by code?

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/UniversalAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,70 @@
+package com.appunite.universal_adapter_kotlin
+
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.ListAdapter
+import com.jacekmarchwicki.universaladapter.BaseAdapterItem
+
+class UniversalAdapter(
```
Comment (jacek-marchwicki) [inline universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/UniversalAdapter.kt:10]: Can't you just use the java version of the UniversalAdapter so two implementations don't need to be maintained?

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt (hunk 1 above)
Comment (marcin-adamczewski) [reply universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt]: true, done

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt (hunk 2 above)
Comment (marcin-adamczewski) [reply universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt]: true, done

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt (hunk 3 above)
Comment (marcin-adamczewski) [reply universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolder.kt]: true, done

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolderManager.kt (hunk 4 above)
Comment (marcin-adamczewski) [reply universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolderManager.kt]: I've never had to create an item view another way, have you? Anyway, I added an abstract method to create the view so it is possible now to take another approach to that.

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/UniversalAdapter.kt (hunk 5 above)
Comment (marcin-adamczewski) [reply universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/UniversalAdapter.kt:10]: That one is tricky. Kotlin version inherits from androidx.ListAdapter which handles list item changes with DiffUtil.
On the other hand, the Java version uses RecyclerView.Adapter and implements ChangesDetector.ChangesAdapter.

There is some common code that could be re-used with some common interface and delegation but the universal-adapter-kotlin would have to depend on universal-adapter module. So we'd unnecessarily have changes detector stuff in the universal-adapter-kotlin module. More to that it's getting complicated. 
Do you have any idea how to approach this?

**Pull request closed without merging**